walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -37,31 +37,76 @@
          return eventsAsTimeRange;
       }
     }
+    // convert events to list representation
+    List<Event> eventsAsList = new ArrayList<Event>();
+    for (Event event: events) {
+      eventsAsList.add(event);
+    }
+    
+    // if there are no mandatory attendees, make optional attendees mandatory attendees
+    if (request.getAttendees().isEmpty()) {
+      List<String> optionalAttendees = new ArrayList<String>();
+      for(String optionalAttendee: request.getOptionalAttendees()) {
+        optionalAttendees.add(optionalAttendee);
+      }
+      MeetingRequest requestOptionalAttendees = new MeetingRequest(optionalAttendees, request.getDuration());
+      List<TimeRange> optionalAttendeesTimeRanges = getTimeRanges(eventsAsList, requestOptionalAttendees, false);
+      return optionalAttendeesTimeRanges;
+    }
 
-    // filter out events that don't involve required attendees
-    List<Event> filteredEvents = filterEvents(events, request);
-
-    if (filteredEvents.isEmpty()){
+    // get time ranges with both required and optional attendees
+    
+    List<TimeRange> allAttendeesTimeRanges = new ArrayList<TimeRange>();
+    if (!(request.getOptionalAttendees().isEmpty())) {
+      // create new MeetingRequest object with both optional and required attendees
+      // must do this ^ because you can't modify the the current attendees
+      List<String> combinedAttendees = new ArrayList<String>();
+      combinedAttendees.addAll(request.getAttendees());
+      combinedAttendees.addAll(request.getOptionalAttendees());
+
+      MeetingRequest requestWithOptionalAttendees = new MeetingRequest(combinedAttendees, request.getDuration());
+      allAttendeesTimeRanges = getTimeRanges(eventsAsList, requestWithOptionalAttendees, true);
+    } 
+
+    // get time ranges with only required attendees
+    List<Event> filteredEvents = filterEvents(eventsAsList, request);  
+
+    // if there are no events and no optional attendees
+    if (filteredEvents.isEmpty() && allAttendeesTimeRanges.isEmpty()){
       eventsAsTimeRange.add(TimeRange.fromStartDuration(0, 1440));
       return eventsAsTimeRange;
     }
+
+    List<TimeRange> requiredAttendeesTimeRanges = new ArrayList<TimeRange>();
+    requiredAttendeesTimeRanges = getTimeRanges(filteredEvents, request, true);
+  
+    // if you can schedule in a meeting with optional attendees, select those
+    // otherwise, return time ranges with only required attendees
+    if (! (allAttendeesTimeRanges.isEmpty())) {
+      return allAttendeesTimeRanges;
+    } else {
+      return requiredAttendeesTimeRanges;
+    }
```

Comment:
simplify logic here. prefer checking the non-negated condition because it's easier to read.

      if (allAttendeesTimeRanges.isEmpty()) {
            return requiredAttendeesTimeRanges;
      } 
      return allAttendeesTimeRanges;